# Keras Exercise

## Predict political party based on votes

As a fun little example, we'll use a public data set of how US congressmen voted on 17 different issues in the year 1984. Let's see if we can figure out their political party based on their votes alone, using a deep neural network!

For those outside the United States, our two main political parties are "Democrat" and "Republican." In modern times they represent progressive and conservative ideologies, respectively.

Politics in 1984 weren't quite as polarized as they are today, but you should still be able to get over 90% accuracy without much trouble.

Since the point of this exercise is implementing neural networks in Keras, I'll help you to load and prepare the data.

Let's start by importing the raw CSV file using Pandas, and make a DataFrame out of it with nice column labels:

In [1]:
import pandas as pd

feature_names =  ['party','handicapped-infants', 'water-project-cost-sharing', 
                    'adoption-of-the-budget-resolution', 'physician-fee-freeze',
                    'el-salvador-aid', 'religious-groups-in-schools',
                    'anti-satellite-test-ban', 'aid-to-nicaraguan-contras',
                    'mx-missle', 'immigration', 'synfuels-corporation-cutback',
                    'education-spending', 'superfund-right-to-sue', 'crime',
                    'duty-free-exports', 'export-administration-act-south-africa']

voting_data = pd.read_csv('house-votes-84.data.txt', na_values=['?'], 
                          names = feature_names)
voting_data.head()

party handicapped-infants water-project-cost-sharing  \
0  republican                   n                          y   
1  republican                   n                          y   
2    democrat                 NaN                          y   
3    democrat                   n                          y   
4    democrat                   y                          y   

  adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
0                                 n                    y               y   
1                                 n                    y               y   
2                                 y                  NaN               y   
3                                 y                    n             NaN   
4                                 y                    n               y   

  religious-groups-in-schools anti-satellite-test-ban  \
0                           y                       n   
1                           y                       n   
2                           y                       n   
3                           y                       n   
4                           y                       n   

  aid-to-nicaraguan-contras mx-missle immigration  \
0                         n         n           y   
1                         n         n           n   
2                         n         n           n   
3                         n         n           n   
4                         n         n           n   

  synfuels-corporation-cutback education-spending superfund-right-to-sue  \
0                          NaN                  y                      y   
1                            n                  y                      y   
2                            y                  n                      y   
3                            y                  n                      y   
4                            y                NaN                      y   

  crime duty-free-exports export-administration-act-south-africa  
0     y                 n                                      y  
1     y                 n                                    NaN  
2     y                 n                                      n  
3     n                 n                                      y  
4     y                 y                                      y

We can use describe() to get a feel of how the data looks in aggregate:

In [3]:
voting_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 435 entries, 0 to 434
Data columns (total 17 columns):
party                                     435 non-null object
handicapped-infants                       423 non-null object
water-project-cost-sharing                387 non-null object
adoption-of-the-budget-resolution         424 non-null object
physician-fee-freeze                      424 non-null object
el-salvador-aid                           420 non-null object
religious-groups-in-schools               424 non-null object
anti-satellite-test-ban                   421 non-null object
aid-to-nicaraguan-contras                 420 non-null object
mx-missle                                 413 non-null object
immigration                               428 non-null object
synfuels-corporation-cutback              414 non-null object
education-spending                        404 non-null object
superfund-right-to-sue                    410 non-null object
crime                      

In [2]:
voting_data.describe()

party handicapped-infants water-project-cost-sharing  \
count        435                 423                        387   
unique         2                   2                          2   
top     democrat                   n                          y   
freq         267                 236                        195   

       adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
count                                424                  424             420   
unique                                 2                    2               2   
top                                    y                    n               y   
freq                                 253                  247             212   

       religious-groups-in-schools anti-satellite-test-ban  \
count                          424                     421   
unique                           2                       2   
top                              y                       y   
freq                           272                     239   

       aid-to-nicaraguan-contras mx-missle immigration  \
count                        420       413         428   
unique                         2         2           2   
top                            y         y           y   
freq                         242       207         216   

       synfuels-corporation-cutback education-spending superfund-right-to-sue  \
count                           414                404                    410   
unique                            2                  2                      2   
top                               n                  n                      y   
freq                            264                233                    209   

       crime duty-free-exports export-administration-act-south-africa  
count    418               407                                    331  
unique     2                 2                                      2  
top        y                 n                                      y  
freq     248               233                                    269

We can see there's some missing data to deal with here; some politicians abstained on some votes, or just weren't present when the vote was taken. We will just drop the rows with missing data to keep it simple, but in practice you'd want to first make sure that doing so didn't introduce any sort of bias into your analysis (if one party abstains more than another, that could be problematic for example.)

In [4]:
voting_data.dropna(inplace=True)
voting_data.describe()

party handicapped-infants water-project-cost-sharing  \
count        232                 232                        232   
unique         2                   2                          2   
top     democrat                   n                          n   
freq         124                 136                        125   

       adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
count                                232                  232             232   
unique                                 2                    2               2   
top                                    y                    n               y   
freq                                 123                  119             128   

       religious-groups-in-schools anti-satellite-test-ban  \
count                          232                     232   
unique                           2                       2   
top                              y                       y   
freq                           149                     124   

       aid-to-nicaraguan-contras mx-missle immigration  \
count                        232       232         232   
unique                         2         2           2   
top                            y         n           y   
freq                         119       119         128   

       synfuels-corporation-cutback education-spending superfund-right-to-sue  \
count                           232                232                    232   
unique                            2                  2                      2   
top                               n                  n                      y   
freq                            152                124                    127   

       crime duty-free-exports export-administration-act-south-africa  
count    232               232                                    232  
unique     2                 2                                      2  
top        y                 n                                      y  
freq     149               146                                    189

Our neural network needs normalized numbers, not strings, to work. So let's replace all the y's and n's with 1's and 0's, and represent the parties as 1's and 0's as well.

In [5]:
r=voting_data

In [6]:
dic={'y':1,'n':0}

In [13]:
r.replace(('y','n'),(1,0),inplace=True)

In [14]:
r.head()

party  handicapped-infants  water-project-cost-sharing  \
5     democrat                    0                           1   
8   republican                    0                           1   
19    democrat                    1                           1   
23    democrat                    1                           1   
25    democrat                    1                           0   

    adoption-of-the-budget-resolution  physician-fee-freeze  el-salvador-aid  \
5                                   1                     0                1   
8                                   0                     1                1   
19                                  1                     0                0   
23                                  1                     0                0   
25                                  1                     0                0   

    religious-groups-in-schools  anti-satellite-test-ban  \
5                             1                        0   
8                             1                        0   
19                            0                        1   
23                            0                        1   
25                            0                        1   

    aid-to-nicaraguan-contras  mx-missle  immigration  \
5                           0          0            0   
8                           0          0            0   
19                          1          1            0   
23                          1          1            0   
25                          1          1            1   

    synfuels-corporation-cutback  education-spending  superfund-right-to-sue  \
5                              0                   0                       1   
8                              0                   1                       1   
19                             1                   0                       0   
23                             0                   0                       0   
25                             0                   0                       0   

    crime  duty-free-exports  export-administration-act-south-africa  
5       1                  1                                       1  
8       1                  0                                       1  
19      0                  1                                       1  
23      0                  1                                       1  
25      0                  1                                       1

In [15]:
voting_data.replace(('y', 'n'), (1, 0), inplace=True)
voting_data.replace(('democrat', 'republican'), (1, 0), inplace=True)

In [16]:
voting_data.head()

party  handicapped-infants  water-project-cost-sharing  \
5       1                    0                           1   
8       0                    0                           1   
19      1                    1                           1   
23      1                    1                           1   
25      1                    1                           0   

    adoption-of-the-budget-resolution  physician-fee-freeze  el-salvador-aid  \
5                                   1                     0                1   
8                                   0                     1                1   
19                                  1                     0                0   
23                                  1                     0                0   
25                                  1                     0                0   

    religious-groups-in-schools  anti-satellite-test-ban  \
5                             1                        0   
8                             1                        0   
19                            0                        1   
23                            0                        1   
25                            0                        1   

    aid-to-nicaraguan-contras  mx-missle  immigration  \
5                           0          0            0   
8                           0          0            0   
19                          1          1            0   
23                          1          1            0   
25                          1          1            1   

    synfuels-corporation-cutback  education-spending  superfund-right-to-sue  \
5                              0                   0                       1   
8                              0                   1                       1   
19                             1                   0                       0   
23                             0                   0                       0   
25                             0                   0                       0   

    crime  duty-free-exports  export-administration-act-south-africa  
5       1                  1                                       1  
8       1                  0                                       1  
19      0                  1                                       1  
23      0                  1                                       1  
25      0                  1                                       1

Finally let's extract the features and labels in the form that Keras will expect:

In [33]:
voting_data.columns

Index(['party', 'handicapped-infants', 'water-project-cost-sharing',
       'adoption-of-the-budget-resolution', 'physician-fee-freeze',
       'el-salvador-aid', 'religious-groups-in-schools',
       'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missle',
       'immigration', 'synfuels-corporation-cutback', 'education-spending',
       'superfund-right-to-sue', 'crime', 'duty-free-exports',
       'export-administration-act-south-africa'],
      dtype='object')

In [37]:
voting_data.head()

party  handicapped-infants  water-project-cost-sharing  \
5       1                    0                           1   
8       0                    0                           1   
19      1                    1                           1   
23      1                    1                           1   
25      1                    1                           0   

    adoption-of-the-budget-resolution  physician-fee-freeze  el-salvador-aid  \
5                                   1                     0                1   
8                                   0                     1                1   
19                                  1                     0                0   
23                                  1                     0                0   
25                                  1                     0                0   

    religious-groups-in-schools  anti-satellite-test-ban  \
5                             1                        0   
8                             1                        0   
19                            0                        1   
23                            0                        1   
25                            0                        1   

    aid-to-nicaraguan-contras  mx-missle  immigration  \
5                           0          0            0   
8                           0          0            0   
19                          1          1            0   
23                          1          1            0   
25                          1          1            1   

    synfuels-corporation-cutback  education-spending  superfund-right-to-sue  \
5                              0                   0                       1   
8                              0                   1                       1   
19                             1                   0                       0   
23                             0                   0                       0   
25                             0                   0                       0   

    crime  duty-free-exports  export-administration-act-south-africa  
5       1                  1                                       1  
8       1                  0                                       1  
19      0                  1                                       1  
23      0                  1                                       1  
25      0                  1                                       1

In [38]:
all_features = voting_data.drop('party',axis=1).values
all_classes = voting_data['party'].values

In [39]:
all_features

array([[0, 1, 1, ..., 1, 1, 1],
       [0, 1, 0, ..., 1, 0, 1],
       [1, 1, 1, ..., 0, 1, 1],
       ..., 
       [0, 0, 0, ..., 1, 0, 1],
       [0, 0, 1, ..., 1, 0, 1],
       [0, 0, 1, ..., 0, 0, 1]], dtype=int64)

In [19]:
all_classes

array([1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0,
       0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0,
       0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0,
       1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1,
       0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0,
       0, 1], dtype=int64)

OK, so have a go at it! You'll want to refer back to the slide on using Keras with binary classification - there are only two parties, so this is a binary problem. This also saves us the hassle of representing classes with "one-hot" format like we had to do with MNIST; our output is just a single 0 or 1 value.

Also refer to the scikit_learn integration slide, and use cross_val_score to evaluate your resulting model with 10-fold cross-validation.

Try out your code here:

In [20]:
import keras

Using TensorFlow backend.


In [21]:
from keras.models import Sequential
from keras.optimizers import RMSprop
from keras.layers import Dense,Dropout

In [24]:
from keras.wrappers.scikit_learn import KerasClassifier

In [25]:
import sklearn

In [26]:
from sklearn.cross_validation import train_test_split

In [40]:
X_train, X_test, y_train, y_test = train_test_split(all_features, all_classes, test_size=0.40, random_state=101)

In [41]:
model = Sequential()

In [42]:
model.add(Dense(64,input_dim=16,activation='relu'))

In [43]:
model.add(Dropout(0.5))
model.add(Dense(64,activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(1,activation='sigmoid'))
model.compile(loss='binary_crossentropy',optimizer='rmsprop',metrics=['accuracy'])

In [44]:
model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_4 (Dense)              (None, 64)                1088      
_________________________________________________________________
dropout_3 (Dropout)          (None, 64)                0         
_________________________________________________________________
dense_5 (Dense)              (None, 64)                4160      
_________________________________________________________________
dropout_4 (Dropout)          (None, 64)                0         
_________________________________________________________________
dense_6 (Dense)              (None, 1)                 65        
Total params: 5,313
Trainable params: 5,313
Non-trainable params: 0
_________________________________________________________________


In [47]:
result=model.fit(X_train,y_train,batch_size=100,epochs=20,verbose=2,validation_data=(X_test,y_test))

Train on 139 samples, validate on 93 samples
Epoch 1/20
0s - loss: 0.1310 - acc: 0.9568 - val_loss: 0.1571 - val_acc: 0.9247
Epoch 2/20
0s - loss: 0.0886 - acc: 0.9784 - val_loss: 0.1576 - val_acc: 0.9247
Epoch 3/20
0s - loss: 0.1197 - acc: 0.9640 - val_loss: 0.1582 - val_acc: 0.9247
Epoch 4/20
0s - loss: 0.1190 - acc: 0.9640 - val_loss: 0.1567 - val_acc: 0.9247
Epoch 5/20
0s - loss: 0.1080 - acc: 0.9568 - val_loss: 0.1560 - val_acc: 0.9247
Epoch 6/20
0s - loss: 0.0859 - acc: 0.9712 - val_loss: 0.1543 - val_acc: 0.9247
Epoch 7/20
0s - loss: 0.0879 - acc: 0.9784 - val_loss: 0.1551 - val_acc: 0.9247
Epoch 8/20
0s - loss: 0.0958 - acc: 0.9712 - val_loss: 0.1542 - val_acc: 0.9247
Epoch 9/20
0s - loss: 0.1180 - acc: 0.9640 - val_loss: 0.1544 - val_acc: 0.9247
Epoch 10/20
0s - loss: 0.1040 - acc: 0.9784 - val_loss: 0.1553 - val_acc: 0.9247
Epoch 11/20
0s - loss: 0.0951 - acc: 0.9712 - val_loss: 0.1544 - val_acc: 0.9247
Epoch 12/20
0s - loss: 0.1371 - acc: 0.9496 - val_loss: 0.1530 - val_acc:

In [48]:
model.evaluate(X_test,y_test,verbose=0)

[0.1492196766919987, 0.93548387801775368]

In [59]:
def create_model():
    model = Sequential()
    model.add(Dense(64,input_dim=16,activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(64,activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(1,activation='sigmoid'))
    model.compile(loss='binary_crossentropy',optimizer='rmsprop',metrics=['accuracy'])
    return model

In [60]:
from sklearn import model_selection

In [61]:
estimator= KerasClassifier(build_fn=create_model,nb_epoch=100,verbose=0)

In [62]:
cv_score =  model_selection.cross_val_score(estimator,all_features, all_classes,cv=10)

In [63]:
print(cv_score.mean())

0.926811593771


## My implementation is below

# No peeking!

![title](peek.jpg)

In [7]:
from keras.layers import Dense, Dropout
from keras.models import Sequential
from sklearn.model_selection import cross_val_score

def create_model():
    model = Sequential()
    #17 feature inputs (votes) going into an 32-unit layer 
    model.add(Dense(32, input_dim=17, kernel_initializer='normal', activation='relu'))
    # Another hidden layer of 16 units
    model.add(Dense(16, kernel_initializer='normal', activation='relu'))
    # Output layer with a binary classification (Democrat or Republican political party)
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='rmsprop', metrics=['accuracy'])
    return model

from keras.wrappers.scikit_learn import KerasClassifier

# Wrap our Keras model in an estimator compatible with scikit_learn
estimator = KerasClassifier(build_fn=create_model, nb_epoch=100, verbose=0)
# Now we can use scikit_learn's cross_val_score to evaluate this model identically to the others
cv_scores = cross_val_score(estimator, all_features, all_classes, cv=10)
cv_scores.mean()

Using TensorFlow backend.


0.93949275612831118

93% without even trying too hard! Did you do better? Maybe more neurons, more layers, or Dropout layers would help even more.